functions/gateway/main.go

```diff
@@ -54,103 +54,174 @@ func init() {
 		{"/api/html/seshu/session/submit", "POST", handlers.SubmitSeshuSession, None},
 		{"/api/html/seshu/session/location", "PATCH", handlers.GeoThenPatchSeshuSession, None},
 		{"/api/html/seshu/session/events", "PATCH", handlers.SubmitSeshuEvents, None},
-    }
+	}
 }
 
-
 type App struct {
-    Router *mux.Router
-     Mw   *authentication.Interceptor[*openid.UserInfoContext[*oidc.IDTokenClaims, *oidc.UserInfo]]
-    AuthN *authentication.Authenticator[*openid.UserInfoContext[*oidc.IDTokenClaims, *oidc.UserInfo]]
+	Router *mux.Router
+	AuthZ  *authorization.Authorizer[*oauth.IntrospectionContext]
 }
 
-
 func NewApp() *App {
-    app := &App{
-        Router: mux.NewRouter(),
-    }
-    log.Printf("App created: %+v", app)
-    app.Router.Use(withContext)
-    app.InitializeAuth()
-    return app
+	app := &App{
+		Router: mux.NewRouter(),
+	}
+	app.Router.Use(withContext)
+	app.InitializeAuth()
+	log.Printf("App created: %+v", app)
+	return app
 }
 
 func (app *App) InitializeAuth() {
-    services.InitAuth()
-    app.Mw, app.AuthN = services.GetAuthMw()
+	services.InitAuth()
+	app.AuthZ = services.GetAuthMw()
 }
 
 func (app *App) SetupRoutes(routes []Route) {
-    for _, route := range routes {
-        app.addRoute(route)
-    }
+	for _, route := range routes {
+		app.addRoute(route)
+	}
 }
 
 func (app *App) addRoute(route Route) {
-    var handler http.HandlerFunc
-    switch route.Auth {
-    case Require:
-        handler = func(w http.ResponseWriter, r *http.Request) {
-            if app.Mw == nil {
-                log.Println("Warning: app.Mw is nil, skipping authentication")
-                route.Handler(w, r).ServeHTTP(w, r)
-                return
-            }
-            app.Mw.RequireAuthentication()(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-                if authentication.IsAuthenticated(r.Context()) {
-                    route.Handler(w, r).ServeHTTP(w, r)
-                } else {
-                    http.Redirect(w, r, "/login", http.StatusFound)
-                }
-
-            })).ServeHTTP(w, r)
-        }
-    case Check:
-        handler = func(w http.ResponseWriter, r *http.Request) {
-            app.Mw.CheckAuthentication()(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-                route.Handler(w, r).ServeHTTP(w, r)
-            })).ServeHTTP(w, r)
-        }
-    default:
-        handler = func(w http.ResponseWriter, r *http.Request) {
-            route.Handler(w, r).ServeHTTP(w, r)
-        }
-    }
-
-    app.Router.HandleFunc(route.Path, handler).Methods(route.Method).Name(route.Path)
-}
+	var handler http.HandlerFunc
+	var accessTokenCookie *http.Cookie
+	var refreshTokenCookie *http.Cookie
+	var err error
+	var refreshTokenCookieErr error
+	switch route.Auth {
+	case Require:
+		handler = func(w http.ResponseWriter, r *http.Request) {
+			accessTokenCookie, err = r.Cookie("access_token")
+			if err != nil {
```

Comment:
Avoid indentation hell
```suggestion
			if err != nil {
			    return 
			}
```